### ISSUE TITLE
encapsulate project to injectable part

### ISSUE THREAD

**Issue by snys98** [skoruba/IdentityServer4.Admin#28]
I'm wondering if it is possible to change the structure of this project to two different project(nuget package).

- **IdentityServer4.Admin**
this project will register all operations to a specific endpoint(like "/identity-server-admin") as a rest api.

- **IdentityServer4.Admin.UI**
this project will register all default page to a specific endpoint(like "/identity-server-admin-ui") and it would be possible to inject some css or js or something.

Just like what swagger&swagger-ui has done.

Besides, thanks for your huge efforts for this repo.

**Comment by skoruba** [comment]
Great idea @snys98.

I think it's absolutely possible. I want to prepare the REST api like you mentioned.

What do you suggest with UI part?

Thank you for your feedback. :)

**Comment by snys98** [comment]
I used to modify [something](https://github.com/microexs/Microex.Swagger) of injecting a UI. 

Here register the middleware which serves the index.html and enable directory browse into the assembly(where we can put pages there):
https://github.com/microexs/Microex.Swagger/blob/8159f6ac84c9672ea938a0f1d760035436641e62/Microex.Swagger/SwaggerUI/SwaggerUIBuilderExtensions.cs#L19

And here to inject things we need to customize the pages.
https://github.com/microexs/Microex.Swagger/blob/8159f6ac84c9672ea938a0f1d760035436641e62/Microex.Swagger/SwaggerUI/SwaggerUIIndexMiddleware.cs#L47
https://github.com/microexs/Microex.Swagger/blob/8159f6ac84c9672ea938a0f1d760035436641e62/Microex.Swagger/SwaggerUI/index.html#L73

But considering there are multiple pages and all these pages requires authentication of a role like "IdentityServerAdmin", it would be much more harder to achieve this.

**Comment by skoruba** [comment]
I am working on REST api, but I am wondering how distribute the set of controllers as nuget package - what do you think about it @snys98? 

I think - it is possible to add into nuget package some repositories and services that are connected with database model but what about api endpoints? :)

**Comment by snys98** [comment]
Maybe we should forget about the MVC pattern but try to implement a middleware instead.

all these could be done with a single line in the pipeline, like:
```
app.UseIdentityAdmin(options=>{
        option.EndPoint = "identity-admin";
        // more configs...
}
```
Then api endpoint("/identity-admin") should deal with all the request that has the url prefix of  "identity-admin" then map the sub-urls to corresponding handler functions.

**Comment by skoruba** [comment]
I've found the solution, that it's possible to add controllers from another assembly like this:

```cs
services.AddMvc()
                .AddApplicationPart(typeof(ClientController).Assembly)
                .AddControllersAsServices();
```

[This post](https://stackoverflow.com/questions/42364745/dynamic-route-prefix-for-controllers-in-separate-library?rq=1) describes how to implement dynamic routing with another assembly.

I've built extension method like this:
```cs
public static IServiceCollection AddIdentityServerAdminApi(this IServiceCollection services, Action<ApiRouteConfiguration> configureRoute)
        {
            var options = new ApiRouteConfiguration();
            services.AddSingleton(options);
            configureRoute?.Invoke(options);

            var serviceProvider = services.BuildServiceProvider();
            var apiRouteConfiguration = serviceProvider.GetService<ApiRouteConfiguration>();

            services.AddMvc()
                .AddApplicationPart(typeof(ClientController).Assembly)
                .AddControllersAsServices()
                .AddApiRoutingConfiguration(apiRouteConfiguration); //based on the SO post

            return services;
        }
```

In `Startup.cs`:

```cs
services.AddIdentityServerAdminApi(configuration =>
            {
                configuration.ApiPrefix = "admin-ui/api/";
            });

```
What do you think about it?

I've wondered about custom middleware and I think this approach above it's better because in custom middleware you are working with "raw" http request and you need implement everything from scratch. :)

**Comment by snys98** [comment]
@skoruba 
That's great. It's actually the way of `Microsoft.AspNetCore.Identity.UI`. I know little about this kind of  implemention. But I guess these may help when dealing with razor pages:
```
https://github.com/aspnet/Identity/blob/9481e88aa1fb3c77c894c5b178099d03979853cd/src/UI/IdentityBuilderUIExtensions.cs#L59
```

**Comment by skoruba** [comment]
@snys98 Great point.  I am working on it but I can't find any solution how to share the resources for localization. Razor class library provides great posibilities how to share UI across among multiple projects, but all available samples are without localization. Any idea? :)

**Comment by skoruba** [comment]
Similiar use case with sharing resx files from class library - https://github.com/aspnet/Localization/issues/328

**Comment by snys98** [comment]
@skoruba 
Sorry, I had a busy time in the last few days. Sounds like you've solved it. Congratulations.

**Comment by skoruba** [comment]
@snys98 Yes, it seems the combination of Razor Class Library and Resources for localization working well.
 
I am currently working on splitting the solution into three parts:
- UI
- EntityFramework
- Services (or Business Logic)
After that I'll create some nuget packages.

Again - thank you for your good ideas.
